File: sdk/javascript/src/services/transform.ts
/**
 * Transform Service
 * Handles text transformation operations
 */

import { AxiosInstance } from 'axios';
import {
  TransformRequest,
  TransformResponse,
  BatchTransformRequest,
  BatchTransformResponse,
  TransformationType,
  TransformOptions
} from '../types';
import { API_ENDPOINTS, MAX_TEXT_LENGTH, DEFAULT_INTENSITY } from '../constants';
import { ValidationError } from '../errors';

export class TransformService {
  constructor(private axios: AxiosInstance) {}

  /**
   * Transform text
   * @param request Transform request
   * @returns Transform response
   * @example
   * ```typescript
   * const result = await client.transform.transform({
   *   text: "This needs to be softer",
   *   transformation_type: "soften",
   *   intensity: 2
   * });
   * ```
   */
  async transform(request: TransformRequest): Promise<TransformResponse> {
    this.validateTransformRequest(request);

    const response = await this.axios.post<TransformResponse>(
      API_ENDPOINTS.TRANSFORM,
      request
    );

    return response.data;
  }

  /**
   * Quick transform with soften
   * @param text Text to soften
   * @param intensity Transformation intensity (0-3)
   */
  async soften(
    text: string,
    intensity: number = DEFAULT_INTENSITY,
    options?: TransformOptions
  ): Promise<TransformResponse> {
    return this.transform({
      text,
      transformation_type: 'soften',
      intensity,
      options
    });
  }

  /**
   * Quick transform with clarify
   * @param text Text to clarify
   * @param intensity Transformation intensity (0-3)
   */
  async clarify(
    text: string,
    intensity: number = DEFAULT_INTENSITY,
    options?: TransformOptions
  ): Promise<TransformResponse> {
    return this.transform({
      text,
      transformation_type: 'clarify',
      intensity,
      options
    });
  }

  /**
   * Quick transform with structure
   * @param text Text to structure
   * @param intensity Transformation intensity (0-3)
   */
  async structure(
    text: string,
    intensity: number = DEFAULT_INTENSITY,
    options?: TransformOptions
  ): Promise<TransformResponse> {
    return this.transform({
      text,
      transformation_type: 'structure',
      intensity,
      options
    });
  }

  /**
   * Quick transform with summarize
   * @param text Text to summarize
   * @param intensity Transformation intensity (0-3)
   */
  async summarize(
    text: string,
    intensity: number = DEFAULT_INTENSITY,
    options?: TransformOptions
  ): Promise<TransformResponse> {
    return this.transform({
      text,
      transformation_type: 'summarize',
      intensity,
      options
    });
  }

  /**
   * Transform terminology (technical to non-technical)
   * @param text Text with technical terms
   */
  async transformTerminology(
    text: string,
    options?: TransformOptions
  ): Promise<TransformResponse> {
    return this.transform({
      text,
      transformation_type: 'terminology',
      intensity: DEFAULT_INTENSITY,
      options
    });
  }

  /**
   * Structure requirements into 4 quadrants
   * @param text Unstructured requirements
   */
  async structureRequirements(
    text: string,
    options?: TransformOptions
  ): Promise<TransformResponse> {
    const response = await this.axios.post<TransformResponse>(
      API_ENDPOINTS.STRUCTURE_REQUIREMENTS,
      {
        text,
        options
      }
    );

    return response.data;
  }

  /**
   * Complete background information
   * @param text Text with potentially missing background
   */
  async completeBackground(
    text: string,
    options?: TransformOptions
  ): Promise<TransformResponse> {
    const response = await this.axios.post<TransformResponse>(
      API_ENDPOINTS.COMPLETE_BACKGROUND,
      {
        text,
        options
      }
    );

    return response.data;
  }

  /**
   * Adjust tone with specific intensity
   * @param text Text to adjust
   * @param intensity Tone adjustment intensity (0-3)
   */
  async adjustTone(
    text: string,
    intensity: number,
    targetTone?: string
  ): Promise<TransformResponse> {
    const response = await this.axios.post<TransformResponse>(
      API_ENDPOINTS.ADJUST_TONE,
      {
        text,
        intensity,
        target_tone: targetTone
      }
    );

    return response.data;
  }

  /**
   * Auto-detect optimal transformation intensity
   * @param text Text to analyze
   * @param transformationType Type of transformation
   */
  async autoDetectIntensity(
    text: string,
    transformationType: TransformationType
  ): Promise<{ intensity: number; confidence: number }> {
    const response = await this.axios.post(
      API_ENDPOINTS.AUTO_DETECT_INTENSITY,
      {
        text,
        transformation_type: transformationType
      }
    );

    return response.data;
  }

  /**
   * Get available tone presets
   */
  async getTonePresets(): Promise<Array<{
    id: string;
    name: string;
    description: string;
    settings: any;
  }>> {
    const response = await this.axios.get(API_ENDPOINTS.TONE_PRESETS);
    return response.data;
  }

  /**
   * Batch transform multiple texts
   * @param request Batch transform request
   */
  async batchTransform(
    request: BatchTransformRequest
  ): Promise<BatchTransformResponse> {
    if (!request.items || request.items.length === 0) {
      throw new ValidationError('No items to transform');
    }

    // Validate each item
    request.items.forEach(item => this.validateTransformRequest(item));

    const response = await this.axios.post<BatchTransformResponse>(
      API_ENDPOINTS.BATCH_TRANSFORM,
      request
    );

    return response.data;
  }

  /**
   * Transform with custom instructions
   * @param text Text to transform
   * @param instructions Custom transformation instructions
   */
  async customTransform(
    text: string,
    instructions: string,
    options?: TransformOptions
  ): Promise<TransformResponse> {
    return this.transform({
      text,
      transformation_type: 'custom',
      intensity: DEFAULT_INTENSITY,
      options: {
        ...options,
        custom_instructions: instructions
      }
    });
  }

  /**
   * Get transformation history
   * @param limit Number of items to retrieve
   * @param offset Offset for pagination
   */
  async getHistory(limit: number = 10, offset: number = 0): Promise<any> {
    const response = await this.axios.get(API_ENDPOINTS.HISTORY, {
      params: { limit, offset }
    });

    return response.data;
  }

  /**
   * Validate transform request
   */
  private validateTransformRequest(request: TransformRequest): void {
    if (!request.text) {
      throw new ValidationError('Text is required');
    }

    if (request.text.length > MAX_TEXT_LENGTH) {
      throw new ValidationError(
        `Text exceeds maximum length of ${MAX_TEXT_LENGTH} characters`
      );
    }

    if (!request.transformation_type) {
      throw new ValidationError('Transformation type is required');
    }

    if (
      request.intensity !== undefined &&
      (request.intensity < 0 || request.intensity > 3)
    ) {
      throw new ValidationError('Intensity must be between 0 and 3');
    }
  }
}

File: sdk/python/tonebridge/services/transform.py
"""
Transform Service
Handles text transformation operations
"""

from typing import Optional, Dict, Any, List, Union
from ..types import (
    TransformationType,
    TransformOptions,
    TransformRequest,
    TransformResponse,
    BatchTransformRequest,
)
from ..constants import (
    API_ENDPOINTS,
    DEFAULT_INTENSITY,
    MAX_TEXT_LENGTH,
)
from ..exceptions import ValidationError


class TransformService:
    """Service for text transformation operations"""
    
    def __init__(self, client):
        """
        Initialize transform service
        
        Args:
            client: ToneBridgeClient instance
        """
        self.client = client
    
    def transform(
        self,
        text: str,
        transformation_type: Union[TransformationType, str],
        intensity: int = DEFAULT_INTENSITY,
        options: Optional[TransformOptions] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Transform text
        
        Args:
            text: Text to transform
            transformation_type: Type of transformation
            intensity: Transformation intensity (0-3)
            options: Transformation options
            metadata: Additional metadata
            
        Returns:
            Transformation response
            
        Example:
            >>> result = client.transform.transform(
            ...     "This needs to be fixed",
            ...     TransformationType.SOFTEN,
            ...     intensity=2
            ... )
        """
        self._validate_transform_request(text, transformation_type, intensity)
        
        request_data = {
            "text": text,
            "transformation_type": transformation_type.value if isinstance(transformation_type, TransformationType) else transformation_type,
            "intensity": intensity,
        }
        
        if options:
            request_data["options"] = options.__dict__ if hasattr(options, '__dict__') else options
        
        if metadata:
            request_data["metadata"] = metadata
        
        return self.client.request("POST", API_ENDPOINTS["TRANSFORM"], json=request_data)
    
    def soften(
        self,
        text: str,
        intensity: int = DEFAULT_INTENSITY,
        options: Optional[TransformOptions] = None,
    ) -> Dict[str, Any]:
        """
        Soften harsh language
        
        Args:
            text: Text to soften
            intensity: Transformation intensity (0-3)
            options: Transformation options
            
        Returns:
            Transformation response
        """
        return self.transform(text, TransformationType.SOFTEN, intensity, options)
    
    def clarify(
        self,
        text: str,
        intensity: int = DEFAULT_INTENSITY,
        options: Optional[TransformOptions] = None,
    ) -> Dict[str, Any]:
        """
        Clarify confusing text
        
        Args:
            text: Text to clarify
            intensity: Transformation intensity (0-3)
            options: Transformation options
            
        Returns:
            Transformation response
        """
        return self.transform(text, TransformationType.CLARIFY, intensity, options)
    
    def structure(
        self,
        text: str,
        intensity: int = DEFAULT_INTENSITY,
        options: Optional[TransformOptions] = None,
    ) -> Dict[str, Any]:
        """
        Structure unorganized text
        
        Args:
            text: Text to structure
            intensity: Transformation intensity (0-3)
            options: Transformation options
            
        Returns:
            Transformation response
        """
        return self.transform(text, TransformationType.STRUCTURE, intensity, options)
    
    def summarize(
        self,
        text: str,
        intensity: int = DEFAULT_INTENSITY,
        options: Optional[TransformOptions] = None,
    ) -> Dict[str, Any]:
        """
        Summarize long text
        
        Args:
            text: Text to summarize
            intensity: Transformation intensity (0-3)
            options: Transformation options
            
        Returns:
            Transformation response
        """
        return self.transform(text, TransformationType.SUMMARIZE, intensity, options)
    
    def transform_terminology(
        self,
        text: str,
        options: Optional[TransformOptions] = None,
    ) -> Dict[str, Any]:
        """
        Transform technical terminology to non-technical
        
        Args:
            text: Text with technical terms
            options: Transformation options
            
        Returns:
            Transformation response
        """
        return self.transform(text, TransformationType.TERMINOLOGY, DEFAULT_INTENSITY, options)
    
    def structure_requirements(
        self,
        text: str,
        options: Optional[TransformOptions] = None,
    ) -> Dict[str, Any]:
        """
        Structure requirements into 4 quadrants
        
        Args:
            text: Unstructured requirements
            options: Transformation options
            
        Returns:
            Structured requirements
        """
        request_data = {"text": text}
        if options:
            request_data["options"] = options.__dict__ if hasattr(options, '__dict__') else options
        
        return self.client.request(
            "POST",
            API_ENDPOINTS["STRUCTURE_REQUIREMENTS"],
            json=request_data
        )
    
    def complete_background(
        self,
        text: str,
        options: Optional[TransformOptions] = None,
    ) -> Dict[str, Any]:
        """
        Complete missing background information
        
        Args:
            text: Text with potentially missing background
            options: Transformation options
            
        Returns:
            Text with completed background
        """
        request_data = {"text": text}
        if options:
            request_data["options"] = options.__dict__ if hasattr(options, '__dict__') else options
        
        return self.client.request(
            "POST",
            API_ENDPOINTS["COMPLETE_BACKGROUND"],
            json=request_data
        )
    
    def adjust_tone(
        self,
        text: str,
        intensity: int,
        target_tone: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Adjust tone with specific intensity
        
        Args:
            text: Text to adjust
            intensity: Tone adjustment intensity (0-3)
            target_tone: Target tone
            
        Returns:
            Tone-adjusted text
        """
        return self.client.request(
            "POST",
            API_ENDPOINTS["ADJUST_TONE"],
            json={
                "text": text,
                "intensity": intensity,
                "target_tone": target_tone,
            }
        )
    
    def auto_detect_intensity(
        self,
        text: str,
        transformation_type: Union[TransformationType, str],
    ) -> Dict[str, Any]:
        """
        Auto-detect optimal transformation intensity
        
        Args:
            text: Text to analyze
            transformation_type: Type of transformation
            
        Returns:
            Optimal intensity and confidence
        """
        return self.client.request(
            "POST",
            API_ENDPOINTS["AUTO_DETECT_INTENSITY"],
            json={
                "text": text,
                "transformation_type": transformation_type.value if isinstance(transformation_type, TransformationType) else transformation_type,
            }
        )
    
    def get_tone_presets(self) -> List[Dict[str, Any]]:
        """
        Get available tone presets
        
        Returns:
            List of tone presets
        """
        return self.client.request("GET", API_ENDPOINTS["TONE_PRESETS"])
    
    def batch_transform(
        self,
        items: List[Dict[str, Any]],
        parallel: bool = True,
        stop_on_error: bool = False,
    ) -> Dict[str, Any]:
        """
        Batch transform multiple texts
        
        Args:
            items: List of transform requests
            parallel: Process in parallel
            stop_on_error: Stop on first error
            
        Returns:
            Batch transformation results
        """
        if not items:
            raise ValidationError("No items to transform")
        
        # Validate each item
        for item in items:
            self._validate_transform_request(
                item.get("text", ""),
                item.get("transformation_type", ""),
                item.get("intensity", DEFAULT_INTENSITY)
            )
        
        return self.client.request(
            "POST",
            API_ENDPOINTS["BATCH_TRANSFORM"],
            json={
                "items": items,
                "parallel": parallel,
                "stop_on_error": stop_on_error,
            }
        )
    
    def custom_transform(
        self,
        text: str,
        instructions: str,
        options: Optional[TransformOptions] = None,
    ) -> Dict[str, Any]:
        """
        Transform with custom instructions
        
        Args:
            text: Text to transform
            instructions: Custom transformation instructions
            options: Transformation options
            
        Returns:
            Custom transformation result
        """
        transform_options = options or TransformOptions()
        transform_options.custom_instructions = instructions
        
        return self.transform(
            text,
            TransformationType.CUSTOM,
            DEFAULT_INTENSITY,
            transform_options
        )
    
    def get_history(
        self,
        limit: int = 10,
        offset: int = 0,
    ) -> Dict[str, Any]:
        """
        Get transformation history
        
        Args:
            limit: Number of items to retrieve
            offset: Offset for pagination
            
        Returns:
            Transformation history
        """
        return self.client.request(
            "GET",
            API_ENDPOINTS["HISTORY"],
            params={"limit": limit, "offset": offset}
        )
    
    def _validate_transform_request(
        self,
        text: str,
        transformation_type: Union[TransformationType, str],
        intensity: int,
    ) -> None:
        """Validate transform request parameters"""
        if not text:
            raise ValidationError("Text is required")
        
        if len(text) > MAX_TEXT_LENGTH:
            raise ValidationError(f"Text exceeds maximum length of {MAX_TEXT_LENGTH} characters")
        
        if not transformation_type:
            raise ValidationError("Transformation type is required")
        
        if intensity < 0 or intensity > 3:
            raise ValidationError("Intensity must be between 0 and 3")